### PR TITLE
[Typescript] Plugin template

### PR DESCRIPTION
**Pull request by soupette**
### What does it do?

It migrates the default plugin generators to TS (we still need to merge them in the typescript folder after #12797 is merged).

It uses esbuild-loader instead of ts-loader in the webpack config.
it also fixes some small bugs.

### Why is it needed?

To have templates in TS for plugins.

### CODE REVIEW

File: packages/core/strapi/lib/commands/builders/admin.js [hunk 1]
```diff
@@ -25,23 +26,27 @@ module.exports = async ({ dir, optimization, forceBuild = true, isTSProject }) =
...
 
   // Always remove the .cache and build folders
-  await strapiAdmin.clean({ dir });
+  // FIXME the BE should remove the build dir and the admin should only
+  // be responsible of removing the .cache dir.
+  await strapiAdmin.clean({ appDir: srcDir, buildDestDir });
 
-  ee({ dir });
+  // @convly shouldn't we use the app dir here?
+  ee({ dir: buildDestDir });
```
Comment (Convly) [inline packages/core/strapi/lib/commands/builders/admin.js:34]: TBC, I'll look at it once we merge our PRs

File: packages/core/admin/package.json [hunk 2]
```diff
@@ -68,6 +68,7 @@
     "css-loader": "6.5.1",
     "date-fns": "2.22.1",
     "dotenv": "8.5.1",
+    "esbuild-loader": "2.18.0",
```
Comment (Convly) [inline packages/core/admin/package.json:71]: Can't you put it in the devDeps?

File: packages/core/admin/index.js [hunk 3]
```diff
@@ -229,13 +121,14 @@ async function watchAdmin({ plugins, dir, host, port, browser, options }) {
...
+    appDir,
     cacheDir,
-    pluginsPath,
     dest,
+    entry,
     env,
-    port,
     options,
+    pluginsPath,
+    // port,
```
Comment (Convly) [inline packages/core/admin/index.js:131]: Why did we remove the port option?

**Review by Convly — CHANGES_REQUESTED** [review]
Some questions

File: examples/kitchensink-ts/src/admin/app.example.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import theme from './extensions/theme';
+
+type Config = {
```
[the remark this answers is not in the document]
Comment (soupette) [reply examples/kitchensink-ts/src/admin/app.example.tsx:3]: Yes exactly! I'll move it soon

File: examples/kitchensink-ts/src/plugins/test/server/bootstrap.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+'use strict';
+
+module.exports = ({ strapi }) => {
```
[the remark this answers is not in the document]
Comment (soupette) [reply examples/kitchensink-ts/src/plugins/test/server/bootstrap.js:3]: Yes for the moment I have not migrated the server folder...

File: packages/core/strapi/lib/commands/watchAdmin.js [hunk 6]
```diff
@@ -1,18 +1,21 @@
...
 const addSlash = require('../utils/addSlash');
+const tsUtils = require('../utils/typescript');
 const strapi = require('../index');
-const getEnabledPlugins = require('../core/loaders/plugins/get-enabled-plugins');
 
 module.exports = async function({ browser }) {
-  const dir = process.cwd();
+  const currentDirectory = process.cwd();
+
+  const isTSProject = await tsUtils.isTypeScriptProject(currentDirectory);
```
[the remark this answers is not in the document]
Comment (soupette) [reply packages/core/strapi/lib/commands/watchAdmin.js:14]: Yes we should! But we need to fix the plugins path before.

File: packages/core/admin/package.json (hunk 2 above)
Comment (soupette) [reply packages/core/admin/package.json:71]: No we need this one at runtime.

File: packages/core/admin/index.js (hunk 3 above)
Comment (soupette) [reply packages/core/admin/index.js:131]: It's an option for the `devServer` and not webpack

**Review by Convly — APPROVED** [review]

**Pull request merged**